Add support for `podman quadlet`

flouthoc opened the pull request:
Introduce support for the `podman quadlet` which allows users to manage and orchestrate containers and objects using `systemd`  using inbuilt `quadlet` support to `podman`  , providing the following subcommands:


<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
Allow users to orchestrate and manage containers using a new podman quadlet command
```

flouthoc commented on cmd/podman/quadlet/install.go:
```diff
@@ -0,0 +1,66 @@
+			}
+			return nil
+		},
+		Example: `podman quadlet install /path/to/myquadlet.container
```
> Furthermore, if you look at the examples, you can see that in many cases, more than one file is needed (units and others). Can the install command support installing a directory?

Yes it supports multiplefile but I think we need to start consuming entire directory as well.

axel7083 replied:
If we copy resources, we may want to ensure they don't already exists, and have a `--force` flag, to allow overwriting existing files

flouthoc replied:
I have added patch to install from directory in new commit.

> If we copy resources, we may want to ensure they don't already exists, and have a --force flag, to allow overwriting existing files

Will add `--force` in a follow up PR once this gets merged in order to keep review simpler and not to added too much to matt's original patch.

flouthoc commented on cmd/podman/quadlet/remove.go:
```diff
@@ -0,0 +1,69 @@
+		return err
+	}
+
+	return nil
```
Fixed.

flouthoc commented:
@mheon Ready for review.

mheon commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,586 @@
+	if err != nil {
+		return nil, fmt.Errorf("querying systemd for unit status: %w", err)
+	}
+	if len(statuses) != len(allServiceNames) {
+		logrus.Warnf("Queried for %d services but received %d responses", len(allServiceNames), len(statuses))
+	}
```
Untrue, systemd will return a response for every entry in the allServiceNames list - if systemd doesn't recognize it, we just get a status report back with a "not loaded/unrecognized" error. If this condition is hit, systemd has done something really wierd and violated its API contract.

Luap99 replied:
Where is the docs for ListUnitsByNames(), I cannot find them.
Because I don't understand how passing in a non existing service (for systemd), i.e. you only added a quadlet file with running daemon-reload makes systemd return a match here. I am clearly missing something here.

flouthoc replied:
It's here https://pkg.go.dev/github.com/coreos/go-systemd/v22/dbus#Conn.ListUnitsByNamesContext

Luap99 commented:
Trying this locally
```
NAME            UNIT NAME   PATH ON DISK                                              STATUS
t1.container                /home/pholzing/.config/containers/systemd/t1.container    inactive/dead
t2.container                /home/pholzing/.config/containers/systemd/t2.container    Not loaded
t5.container                /home/pholzing/.config/containers/systemd/t5.container    Not loaded
test.container              /home/pholzing/.config/containers/systemd/test.container  inactive/dead
```

UNIT NAME is always empty, that does not seem right?

flouthoc commented on cmd/podman/quadlet/install.go:
```diff
@@ -0,0 +1,68 @@
+
+	if len(installReport.QuadletErrors) > 0 {
+		return errors.New("errors occurred installing some Quadlets")
+	}
```
Done.

flouthoc commented on cmd/podman/quadlet/remove.go:
```diff
@@ -0,0 +1,64 @@
+
+		if err == nil && len(removeReport.Errors) > 0 {
+			return errors.New("some quadlets could not be removed")
+		}
```
Done.

flouthoc commented on cmd/quadlet/main.go:
```diff
@@ -580,33 +364,32 @@ func generateUnitsInfoMap(units []*parser.UnitFile) map[string]*quadlet.UnitInfo
+		case strings.HasSuffix(unit.Filename, ".volume"), strings.HasSuffix(unit.Filename, ".kube"):
+			resourceName = ""
+		case strings.HasSuffix(unit.Filename, ".network"), strings.HasSuffix(unit.Filename, ".image"):
+			resourceName = ""
```
It's because some part of code was being used elsewhere to it was moved to a different package and if you see in previous version `resourceName` was not set for these cases, I have merged it into a single case I hope it explains better.

flouthoc commented on docs/source/markdown/podman-quadlet-install.1.md:
````diff
@@ -0,0 +1,43 @@
+Install quadlet from artifact
+```
+$ podman quadlet install oci-artifact://my-artifact:latest
+```
````
Removed.

flouthoc commented on test/e2e/quadlet/alpine-quadlet.container:
```diff
@@ -0,0 +1,11 @@
+Exec=sleep infinity
+
+[Service]
+#Restart=always
```
Removed.

flouthoc commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,586 @@
+		quadletRootlessDirs := systemdquadlet.GetUnitDirs(true)
+
+		for _, dir := range quadletRootlessDirs {
+			// Prefer /etc/containers/systemd/users(/$UID)
```
Yes if we can have consensus among maintainers on certain paths then I can do the required changes.

ygalblum replied:
I don't think there should be an order at all. There should be one default target overridable by the caller.

The defaults should be `/etc/containers/systemd` for `root` and `os.UserConfigDir()/containers/systemd` for `rootless`. The user should be able to override it with an optional argument (say `--install-dir`).

There's already a discussion on directories, so they should be copied as is into these targets (that is staying in a separate directory). Not sure if we discussed tarballs and such. But, that's another good way to pass multiple files. So, they should be extracted into the target directory

flouthoc replied:
I like this idea. @mheon @Luap99 any thoughts ?

Luap99 replied:
yes exactly, this needs to be one path per user where we install to

flouthoc commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,586 @@
+	for _, toInstall := range pathsOrURLs {
+		switch {
+		case strings.HasPrefix(toInstall, "oci-artifact://"):
+			artName := strings.TrimPrefix(toInstall, "oci-artifact://")
```
Done.

flouthoc commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,586 @@
+					} else {
+						installReport.InstalledQuadlets[toInstall] = installedPath
+					}
+				}
```
Done.

flouthoc commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,586 @@
+		return func(q *entities.ListQuadlet) bool {
+			var filters []string
+			for _, f := range filterValues {
+				filters = append(filters, strings.ReplaceAll(f, "/", ""))
```
Removed, I copied it from another filter implementation.

flouthoc commented on test/e2e/quadlet_verb_test.go:
```diff
@@ -0,0 +1,59 @@
+		SkipIfSystemdNotRunning("cannot test systemd is not running")
+	})
+
+	It("podman quadlet install, list, rm", func() {
```
I added systems test.

flouthoc commented:
> Trying this locally
> 
> ```
> NAME            UNIT NAME   PATH ON DISK                                              STATUS
> t1.container                /home/pholzing/.config/containers/systemd/t1.container    inactive/dead
> t2.container                /home/pholzing/.config/containers/systemd/t2.container    Not loaded
> t5.container                /home/pholzing/.config/containers/systemd/t5.container    Not loaded
> test.container              /home/pholzing/.config/containers/systemd/test.container  inactive/dead
> ```
> 
> UNIT NAME is always empty, that does not seem right?

Fixed in latest commit, also addressed most of the comments. Remaining issues which are still not solved

* Predictable paths for quadlet files for `rootless` and `root` users. (Waiting for @ygalblum opinion on this.)
* Still bloated, not sure if I got log package correctly.

flouthoc commented:
> One general issue that keep repeating in my comments is about non-Quadlet files. Not supporting them creates an issue

We don't support `non-quadlet` files in this current PR, however if we want to support non-quadlet files, we must come up with a way so `podman` can associate `non-quadlet` files to a `quadlet` file so whenever a `remove` is done we can remove `non-quadlet` files along with `quadlet` files.

I propose a method like this `podman quadlet install myquadlet.container:/associated/non-quadlets/ifany,/second/non-quadlet-path` , and podman can maintain a list of all non quadlet files in a hidden file called `.myquadlet.container.assets` to store list of `non-quadlet` files associated to a quadlet and can remove those if needed.

@ygalblum @Luap99 @mheon WDYT ?

mheon commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,555 @@
+	// really has to parse everything), but we should be able to output in a machine-readable format
+	// (JSON, maybe?) so we can easily associated error to quadlet here, and return better
+	// results.
+	var validateErr error
```
Does that work if the quadlet has dependencies on already-installed quadlets?

mheon commented:
Using a hidden file SGTM, but that user interface seems pretty bad. Should we restrict `podman quadlet install` to a single Quadlet at a time, the first argument is the quadlet, all subsequent arguments are extra files to install with the quadlet?

ygalblum commented:
Installing a single Quadlet file will generate errors in daemon reload if other units are missing. For example, if you install the `.container` file that depends on a `.network` file that wasn't yet installed, the processing of the former will fail. Then you start mandating not only multiple calls, but also that the call order.
Maybe a simpler approach is to support either single units or directories.

flouthoc commented:
> Installing a single Quadlet file will generate errors in daemon reload if other units are missing. For example, if you install the `.container` file that depends on a `.network` file that wasn't yet installed, the processing of the former will fail. Then you start mandating not only multiple calls, but also that the call order. Maybe a simpler approach is to support either single units or directories.

Problem with directory is, if a directory contains two `quadlet` files and contains different `non-quadlet` files, then how will `podman` know which `non-quadlet` file belongs which `quadlet` file ?

ygalblum commented:
> Or in case of a directory we can just associate all non-quadlet file to one quadlet file in directory and assume users to be aware that mutiple quadlet files should be orchestrated or removed together.

Maybe when it's a directory, the identifier shouldn't be at a Quadlet unit level, but an application or deployment (find better name) level.

flouthoc commented:
> > Or in case of a directory we can just associate all non-quadlet file to one quadlet file in directory and assume users to be aware that mutiple quadlet files should be orchestrated or removed together.
> 
> Maybe when it's a directory, the identifier shouldn't be at a Quadlet unit level, but an application or deployment (find better name) level.

I think quadlet list should still show all quadlets independentaly but with another column if they belong to common app and deleting a quadlet from app, should remove entire `app` ?

So it both should remove app

* `quadlet remove unit` --> if belongs to app, removes entire app
* `quadlet remove app` --> also removes app.

flouthoc commented on test/system/252-quadlet.bats:
```diff
@@ -1865,4 +1865,164 @@ EOF
     run_podman rmi -i $image_tag
 }
+
+@test "quadlet verb - install, list, rm" {
```
I thought `quadlet` and `quadlet verb` are togather part of similar big feature, since helpers functions are being shared therefore I added them in same file. These are only few new tests and I have added `quadlet verb` as suffix to differentiate them.

flouthoc commented:
@mheon @Luap99 @ygalblum This is ready for review.

flouthoc commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,555 @@
+						continue
+					}
+					for _, file := range files {
+						if file.IsDir() {
```
Fixed.

flouthoc commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,555 @@
+	return string(contents), nil
+}
+
+func (ic *ContainerEngine) QuadletRemove(ctx context.Context, quadlets []string, options entities.QuadletRemoveOptions) (*entities.QuadletRemoveReport, error) {
```
Added this feature and also added test.

ygalblum commented on test/system/252-quadlet.bats:
```diff
@@ -1865,4 +1865,312 @@ EOF
+    assert "$output" =~ "Mount=type=bind,source=./test.txt,destination=/test.txt:z" "print should show volume mount configuration"
+
+    # Test quadlet rm
+    run_podman quadlet rm mount-test.container
```
Is this call expected to also remove the "$install_dir/test.txt"? If yes, it should be verified. If not, aren't we leaving leftovers?

ygalblum commented on test/system/252-quadlet.bats:
```diff
@@ -1865,4 +1865,312 @@ EOF
+LogDriver=passthrough
+EOF
+    # Clean all quadlets
+    run_podman quadlet rm --all -f
```
If I run this locally, it will remove all my Quadlet files.

ygalblum commented on test/system/252-quadlet.bats:
```diff
@@ -1865,4 +1865,312 @@ EOF
+# bats test_tags=distro-integration
+@test "quadlet verb - install multiple quadlets from directory and remove by application" {
+    # Create a temporary directory for multiple quadlet files
+    local app_name="test-app-$(random_string)"
```
Use `safe_name` instead of `random_string` (throughout)

ygalblum commented on test/system/252-quadlet.bats:
```diff
@@ -1865,4 +1865,312 @@ EOF
+# bats test_tags=distro-integration
+@test "quadlet verb - install multiple files from directory" {
+    # Create a directory for multiple quadlet files
+    local quadlet_dir=$PODMAN_TMPDIR/quadlet-multi
```
Always use `safe_name` to allow parallelism

ygalblum reviewed (commented):
I still need to review the `abi`. But, I wanted to put in these comments

mheon commented on docs/source/markdown/podman-quadlet-install.1.md:
```diff
@@ -0,0 +1,44 @@
+
+## DESCRIPTION
+
+Install a Quadlet file or an application (which may include multiple Quadlet files) for the current user. You can specify Quadlet files as local files or web URLs.
```
Need more details on how applications work here - it's not very clear

mheon commented on docs/source/markdown/podman-quadlet-list.1.md:
```diff
@@ -0,0 +1,55 @@
+
+| Filter     | Description                                                                                      |
+|------------|--------------------------------------------------------------------------------------------------|
+| name       | Filter by quadlet name.                                                                          |
```
Can you add a TODO somewhere to add an Application filter to this?

flouthoc replied:
Added in code.

mheon commented on docs/source/markdown/podman-quadlet.1.md:
```diff
@@ -0,0 +1,25 @@
+| install | [podman-quadlet-install(1)](podman-quadlet-install.1.md)   | Install a quadlet file or quadlet application                |
+| list    | [podman-quadlet-list(1)](podman-quadlet-list.1.md)         | List installed quadlets                                      |
+| print   | [podman-quadlet-print(1)](podman-quadlet-print.1.md)       | Display the contents of a quadlet                            |
+| rm      | [podman-quadlet-rm(1)](podman-quadlet-rm.1.md)             | Removes an installed quadlet                                 |
```
I feel like at some point we need a `quadlet inspect` that prints more details. Particularly about the app the quadlet is part of. But that is followon work.

mheon commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,708 @@
+		quadlet := quadlets[i]
+		quadletPath, err := getQuadletByName(quadlet)
+		if options.All {
+			quadletPath = quadlet
```
This doesn't seem right. Why are we even doing `getQuadletByName` when `options.All` is set?

Luap99 commented:
I haven't found the time for a proper review, I try to get it done next week.
quadlet still shows 66457 size increase which seems unexpected form a PR that should not alter the quadlet binary behavior in any meaningful way.

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,699 @@
+		dents, err := os.ReadDir(dir)
+		if err != nil {
+			// This is perfectly normal, some quadlet directories aren't created by the package
+			logrus.Infof("Cannot list Quadlet directory %s: %v", dir, err)
```
if it is normal then there is no reason to confuse users with errors if they run with log level info?
I would say ErrNotExist should be ignored but other errors definly should be logged then on warning level.

Luap99 reviewed (changes requested):
not a complete review, I still have to look at the test and changed quadlet package but I should be done with most code

Luap99 commented on test/system/253-podman-quadlet.bats:
```diff
@@ -0,0 +1,323 @@
+    assert "$output" !~ "alpine-quadlet.container" "list should not contain removed container"
+}
+
+# bats test_tags=distro-integration
```
please drop these tags, nothing in these tests depends in any way on an actual external dependency so there is no real reason to specially run them in our fedora openQA env.

Luap99 reviewed (changes requested):
Missing tests for --reload-systemd on both install and remove.
Missing tests for --format on list

Missing test for rm --force, and in general you never run/start a quadlet so the force part doesn't seem to be exercised at all.
rm --all also doesn't seem to be actually tested other than the fact it is used for the inital cleanup but we never check if it actually works there.
--ignore is not tested either

And lastly there seem to be no output check on either install and remove commands, there should be exact matches so we now we don't print unwanted extra stuff.

Luap99 commented on pkg/systemd/quadlet/unitdirs.go:
```diff
@@ -0,0 +1,248 @@
+	if rootless {
+		configDir, err := os.UserConfigDir()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Warning: %v", err)
```
this should not print directly to stderr. if we have the logiface then just use this.

Luap99 commented on pkg/logiface/logiface.go:
```diff
@@ -0,0 +1,24 @@
+package logiface
+
+type Logger interface {
+	Errorf(format string, args ...interface{})
```
small nit but modern code should use `any` over `interface{}` same in the other cases here.

flouthoc replied:
I will do this in a followup PR, left it because it was `nit`.

flouthoc commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,699 @@
+	}
+
+	// Is Quadlet installed? No point if not.
+	quadletPath := "/usr/lib/systemd/system-generators/podman-system-generator"
```
`FindHelperBinary` returns empty path for `podman-system-generator`

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,699 @@
+				installReport.QuadletErrors[toInstall] = fmt.Errorf("populating temporary file: %w", err)
+				continue
+			}
+			installedPath, err := ic.installQuadlet(ctx, tmpFile.Name(), "", installDir, assetFile, validateQuadletFile)
```
Isn't this is completely backwards? IF I have a URL xxx/blah.container then I want that thing names blah.container and not anything else? Using a random name just makes this not deterministic which is horrible.

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,699 @@
+}
+
+// Extracts file extension from Content-Disposition or URL
+func getFileExtension(resp *http.Response) string {
+	// Try Content-Disposition header first
+	cd := resp.Header.Get("Content-Disposition")
+	if cd != "" {
+		re := regexp.MustCompile(`(?i)filename="?([^"]+)"?`)
+		matches := re.FindStringSubmatch(cd)
+		if len(matches) > 1 {
+			return path.Ext(matches[1])
+		}
+	}
+	// Fallback: use URL path
+	ext := path.Ext(resp.Request.URL.Path)
+	return ext
```
In what cases is there no filename in the URL? If this is needed that it needs tests as well!
Also not a fan of using regex to parse a simple header and from looking around there are already funcitons to handle parsing in the std: https://pkg.go.dev/mime#ParseMediaType

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,764 @@
+	installDir := systemdquadlet.GetInstallUnitDirPath(rootless.IsRootless())
+	logrus.Debugf("Going to install Quadlet to directory %s", installDir)
+
+	_, err = os.Stat(installDir)
+	// Make the directory if it doesn't exist
+	if err != nil && errors.Is(err, fs.ErrNotExist) {
+		if err := os.MkdirAll(installDir, 0755); err != nil {
+			return nil, fmt.Errorf("unable to create Quadlet install path %s: %w", installDir, err)
+		}
```
you don't need to stat at all, MkdirAll() alone is enough as it doesn't error when the dir already exists

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,764 @@
+		if err := os.MkdirAll(installDir, 0755); err != nil {
+			return nil, fmt.Errorf("unable to create Quadlet install path %s: %w", installDir, err)
+		}
+	} else if err != nil {
+		return nil, fmt.Errorf("unable to stat Quadlet install path %s: %w", installDir, err)
+	}
```
the way these branches are written `err != nil` is checked twice which makes no sense and that is very confusing way to write this. the  `errors.Is(err, fs.ErrNotExist)` should be inside the first branch but see the previous comment this isn't needed at all.

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,764 @@
+//go:build !remote
+// +build !remote
```
old build tag format should still be dropped

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,764 @@
+
+	assetFile := ""
+	paths := pathsOrURLs
+	if len(pathsOrURLs) > 0 && !strings.HasPrefix(pathsOrURLs[0], "http://") && !strings.HasPrefix(pathsOrURLs[0], "https://") {
```
I am still missing an explanation why we special code the first arg to only allow directories. Why do we not just loop over all args and treat every arg the same.

flouthoc replied:
It's because we are identifying a install as `app` if first path is `directory` then we treat all other files in arguments as part of app otherwise we treat all arguments as isolated quadlets.

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,764 @@
+			if err != nil {
+				return nil, fmt.Errorf("unable to read Quadlet dir %s: %w", pathsOrURLs[0], err)
+			}
+			redoPaths := []string{}
```
this can preallocate the slice capacity to avoid so many memory copies
`make([]string, 0, len(entries))`

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,764 @@
+				installReport.QuadletErrors[toInstall] = fmt.Errorf("populating temporary file: %w", err)
+				continue
+			}
+			installedPath, err := ic.installQuadlet(ctx, tmpFile.Name(), "", installDir, assetFile, validateQuadletFile)
```
Again this makes no sense,
```
podman quadlet install https://github.com/containers/podman/blob/main/test/e2e/quadlet/basic.container
/home/pholzing/.config/containers/systemd/quadlet-dl-3360325579.container
```

This results in a quadlet with a completely random name which is just nonsense, it should take the filename of the URL. That is after all why installQuadlet() has the third argument to specify the name we want.

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,764 @@
+	// Try Content-Disposition header first
+	cd := resp.Header.Get("Content-Disposition")
+	if cd != "" {
+		re := regexp.MustCompile(`(?i)filename="?([^"]+)"?`)
```
see my previous comment about this, https://github.com/containers/podman/pull/26330#discussion_r2182934570 which has a well defined function on how to parse this header

Also then return the full filename so the caller can use this as final quadlet name at least.

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,764 @@
+		return "", fmt.Errorf("%q is not a supported Quadlet file type", filepath.Ext(finalPath))
+	}
+
+	// Second, validate that the dest path does NOT exist.
+	// TODO: Overwrite flag?
+	if _, err := os.Stat(finalPath); err == nil {
+		return "", fmt.Errorf("a Quadlet with name %s already exists, refusing to overwrite", filepath.Base(finalPath))
+	}
```
the stat is not needed, from the OpenFile() check for ErrExist error and then return this message

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,764 @@
+
+	file, err := os.OpenFile(finalPath, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0644)
+	if err != nil {
+		return "", fmt.Errorf("unable to open file %s: %v", filepath.Base(finalPath), err)
+	}
+	defer file.Close()
+
+	// Move the file in
+	srcFile, err := os.Open(path)
+	if err != nil {
+		return "", fmt.Errorf("unable to open file %s: %v", path, err)
```
the error messages "stutter" since err alreayd includes the full path you are adding it again so the path will be there twice. Also again here is the %v wrapping errors, it must use %w.

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,764 @@
+	appMap := make(map[string][]string)
+
+	err := filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
```
this can already fail with ErrNotExist so this must be ignored here too.
`podman quadlet list` is completely broken right now when the `~/.config/containers/systemd/` doesn't exist because it fails to builds the appMap here

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,764 @@
+//   - Reloads systemd daemon if necessary
+//
+// Parameters:
+//   - ctx: Context for cancellation and timeout handling
```
it can be assumed that people know what a context is documenting this is just pointless noise.

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,764 @@
+	return string(contents), nil
+}
+
+// QuadletRemove removes one or more Quadlet files from the system.
+// It handles both individual Quadlet files and Quadlet applications (collections of related files).
+// The function performs the following operations:
+//   - Validates that specified Quadlet files exist
+//   - Checks if associated systemd services are running
+//   - Optionally stops running services (if force option is set)
+//   - Removes the Quadlet files and any associated asset files
+//   - Reloads systemd daemon if necessary
+//
+// Parameters:
+//   - ctx: Context for cancellation and timeout handling
+//   - quadlets: List of Quadlet names to remove, or empty slice if options.All is true
+//   - options: Configuration options including Force, Ignore, and All flags
+//
+// Returns:
+//   - QuadletRemoveReport: Contains lists of successfully removed quadlets and any errors encountered
+//   - error: Non-nil if a fatal error occurred that prevented the operation from completing
+//
+// The function supports:
+//   - Force removal of running Quadlet services
+//   - Ignoring missing Quadlet files (with Ignore option)
+//   - Removing all installed Quadlets (with All option)
+//   - Handling Quadlet applications (.app files containing multiple related Quadlets)
```
Personally not sure how I feel about all this AI text, I mean it isn't exactly wrong but it is just such a wall of text with very little actual important information.

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,764 @@
+	if err != nil {
+		return nil, fmt.Errorf("unable to build app map: %w", err)
+	}
+	quadletsRe := []string{}
```
The variable is non descriptive I have no idea what it store from the name

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,764 @@
+	serviceNameToQuadletName := make(map[string]string)
+	needReload := false
+
+	// Early escape: if 0 quadlets are requested, bail immediately without error.
```
The comment is just straight up wrong. We error out not without error. And I mean the comment text itself doesn't add value because it is clear from the code what it does. code comments need to say why not what unless the what itself is hard to follow/understand.

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,764 @@
+		quadlets = allQuadlets
+	}
+
+	for i := 0; i < len(quadlets); i++ {
```
why is this using an index loop, we should be using for range in go for a simple string slice iteration.

Luap99 replied:
It is extremely hard to reason about this entire loop block. With something like this there is the obvious question of how we ensure we do not loop forever if we keep adding entries.
Yes I cannot see that this is broken but it does take quite some work for to actual parse and understand that block there.
Maybe this can be written in a more readable way that does not require such in loop modifications

flouthoc replied:
Can i refactor this later ? In a followup PR.

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,555 @@
+	}
+
+	// Reload systemd, if necessary/requested.
+	if needReload {
```
yes this comment is still open, ReloadSystemd is ignored

Luap99 commented on pkg/systemd/quadlet/unitdirs.go:
```diff
@@ -0,0 +1,247 @@
+
+	for _, eachUnitDir := range strings.Split(unitDirsEnv, ":") {
+		if !filepath.IsAbs(eachUnitDir) {
+			logiface.Errorf("%s not a valid file path", eachUnitDir)
```
logiface is never initialized for quadlet which means all the log lines are now silenced for quadlet which is wrong.

flouthoc commented:
@Luap99 PTAL

ygalblum commented on test/system/253-podman-quadlet.bats:
```diff
@@ -0,0 +1,552 @@
+
+    basic_setup
+    # Clean all quadlets
+    run_podman quadlet rm --all -f
```
I don't like this. This means that if I run this on my local machine all my local Quadlet files will be erased. How can we prevent this from happening or at least alert the developer about the consequences

ygalblum commented on test/system/253-podman-quadlet.bats:
```diff
@@ -0,0 +1,552 @@
+}
+
+function teardown() {
+    for UNIT_FILE in ${UNIT_FILES[@]}; do
```
Maybe I'm missing something, but I don't see where values are added to `UNIT_FILES`

flouthoc replied:
This part is not needed.

flouthoc commented:
@Luap99 PTAL

Luap99 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -0,0 +1,748 @@
+	if err != nil {
+		return nil, fmt.Errorf("unable to build app map: %w", err)
+	}
+	// Is systemd available to the current user?
+	// We cannot proceed if not.
+	conn, err := dbus.NewWithContext(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("connecting to systemd dbus: %w", err)
+	}
```
That is why the list test is broken, you are running against the root socket all the time, this must use `conn, err := systemd.ConnectToDBUS()` like you did in install and remove.

flouthoc commented:
@Luap99 PTAL again.

Luap99 reviewed (approved):
LGTM
